pikelet/src/surface/lex.rs

```diff
@@ -0,0 +1,190 @@
+
+/// The complete set of `LexToken`s some of which never escape the lexer.
+/// See Token for a list of which Tokens do and do not escape.
+#[derive(Logos)]
+enum LexToken {
+    #[end]
+    EOF,
+    #[token = ":"]
+    Colon,
+    #[token = ","]
+    Comma,
+    #[token = "fun"]
+    Fun,
+    #[token = "=>"]
+    DArrow,
+    #[token = "->"]
+    Arrow,
+    #[token = "("]
+    LParen,
+    #[token = ")"]
+    RParen,
+    #[token = "["]
+    LBrack,
+    #[token = "]"]
+    RBrack,
+    #[token = "{"]
+    LBrace,
+    #[token = "}"]
+    RBrace,
+    #[token = r"record"]
+    RecordTerm,
+    #[token = r"Record"]
+    RecordType,
+    #[token = "."]
+    Dot,
+    #[token = "="]
+    Equal,
+    // Hmm, not sure why this doesn't work.
+    // #[regex = r#"'(.|\\"|\\')*'"#]
+    #[regex = r#"('[^'\\]|\\t|\\u|\\n|\\"|\\')*'"#]
+    CharLit,
+    // Ditto.
+    // #[regex = r#""(.|\\"|\\')*""#]
+    #[regex = r#""([^"\\]|\\t|\\u|\\n|\\"|\\')*""#]
+    StrLit,
+    #[regex = r"[-+]?[0-9]+(\.[0-9]+)?"]
+    Number,
+    #[regex = r"[a-zA-Z][a-zA-Z0-9\-]*"]
+    Name,
+    #[regex = r"\^[0-9]+(\.[0-9]+)?"]
+    Shift,
+    #[regex = r"\p{Whitespace}"]
+    Whitespace,
+    #[error]
+    Error,
+}
+
+/// The subset of `LexToken`s for the parser.
+/// The tokens in `LexToken`s which are excluded from this enum are:
+/// * Whitespace -- skipped.
+/// * EOF -- turned into None, rather than a token.
+/// * Error -- turned into Some(Err(...)).
+#[derive(Debug, Clone)]
+pub enum Token<'a> {
+    Colon,
+    Comma,
+    Fun,
+    DArrow,
+    Arrow,
+    LParen,
+    RParen,
+    LBrack,
+    RBrack,
+    LBrace,
+    RBrace,
+    RecordTerm,
+    RecordType,
+    Equal,
+    Dot,
+    CharLit(&'a str),
+    StrLit(&'a str),
+    Number(&'a str),
+    Name(&'a str),
+    Shift(&'a str),
+}
+
+impl<'a> Display for Token<'a> {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        use Token as T;
+        match self {
+            T::Colon => write!(f, ":"),
+            T::Comma => write!(f, ","),
+            T::Fun => write!(f, "Fun"),
+            T::DArrow => write!(f, "=>"),
+            T::Arrow => write!(f, "->"),
+            T::LParen => write!(f, "("),
+            T::RParen => write!(f, ")"),
+            T::LBrack => write!(f, "["),
+            T::RBrack => write!(f, "]"),
+            T::LBrace => write!(f, "{{"),
+            T::RBrace => write!(f, "}}"),
+            T::RecordTerm => write!(f, "record"),
+            T::RecordType => write!(f, "Record"),
+            T::Equal => write!(f, "="),
+            T::Dot => write!(f, "."),
+            T::CharLit(s) => write!(f, "CharLit({})", s),
+            T::StrLit(s) => write!(f, "StrLit({})", s),
+            T::Number(s) => write!(f, "Number({})", s),
+            T::Name(s) => write!(f, "Write({})", s),
+            T::Shift(s) => write!(f, "Shift({})", s),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct LexicalError(std::ops::Range<usize>, &'static str);
```

Comment:
What kind of error do we get from Logos, and how easy is it to customise/match on? And how well does it support error recovery? Ie. continuing to try to produce tokens even if it fails on one?

Reply:
The error we get is AFAICT just a normal token, not even an Error.
So it basically gives you an enum element, which has a range, and slice.
Then you convert that into an error yourself, (At least I think thats how it works).

I added a test for error recovery behavior, and yes it does.